File: 46 trapping rain water/TranWater.py
class Solution(object):
    # Two pointers: Space O(1) Time O(n)
    def trap1(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        n = len(height)
        if n <= 2:
            return 0
        left_max = height[0]
        right_max = height[n - 1]
        
        left = 0
        right = n - 1
        
        res = 0
        
        while (left < right):
            if height[left] < height[right]:
                if left_max < height[left]:
                    left_max = height[left]
                elif left_max > height[left]:
                    res += left_max - height[left]
                left += 1
            else:
                if right_max < height[right]:
                    right_max = height[right]
                elif right_max > height[right]:
                    res += right_max - height[right]
                right -= 1
        
        return res
    
    # Dynamic programming: Space O(n) Time O(n)
    def trap2(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        n = len(height)
        if n <= 2:
            return 0
        left_max = [0] * n
        right_max = [0] * n
        left_max[0] = height[0]
        right_max[n - 1] = height[n - 1]
        
        for i in range(1, n - 1):
            left_max[i] = max(left_max[i - 1], height[i])
        
        for i in range(n - 2, 0, -1):
            right_max[i] = max(right_max[i + 1], height[i])
        
        res = 0
        
        for i in range(1, n - 1):
            water = min(left_max[i], right_max[i]) - height[i]
            if water < 0:
                continue
            res += water
        
        return res
    
    # Monolithic Stack: Time O(n) Space O(n)
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        stack, res = [], 0
        
        for i, v in enumerate(height):
            while len(stack) > 0 and height[stack[-1]] < v:
                cur = height[stack.pop()]
                if len(stack) == 0:
                    break
                res += (min(height[stack[-1]], v) - cur) * (i - stack[-1] - 1)
            
            stack.append(i)      
        
        return res

File: 46 trapping rain water/TranWater.java


public class TranWater {
    /**
     *  Time: O(n) Space: O(1)
     *  leftMax: represents max value before left pointer (left)
     *  rightMax: represents max value after right pointer (right)
     *  each time we compare height at left pointer and height at right pointer.
     *  if height[left] < height[right]: means that how many water this cell (pointed by left)
     *  can trap depends on max(height[left], leftMax). if leftMax <= height[left], no water 
     *  can be trapped in this cell, we update leftMax. if leftMax > height[left], we can 
     *  trap (leftMax - height[left]) water in this cell.
     *  
     *  if height[right] >= height[right]: we use rightMax to do the same thing as above.
     */
    public int trap1(int[] height) {
        if (height == null || height.length <= 2) {
            return 0;
        }
        int leftMax = height[0];
        int rightMax = height[height.length - 1];
        int left = 0;
        int right = height.length - 1;
        int res = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] > leftMax) {
                    leftMax = height[left];
                } else if (height[left] < leftMax) {
                    res += leftMax - height[left];
                }
                left++;
            } else {
                if (height[right] > rightMax) {
                    rightMax = height[right];
                } else if (height[right] < rightMax) {
                    res += rightMax - height[right];
                }
                right--;
            }
        }
        return res;
    }
    /**
     * 
     * Dynamic Programming.
     * Time: O(n) Space: O(n)
     * we use int[] leftMax and int[] rightMax to store leftMax value and rightMax value for each
     * position. And add them as result.
     * 
     */
    public int trap2(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }
        int n = height.length;
        int[] leftMax = new int[n];
        leftMax[0] = height[0];
        int[] rightMax = new int[n];
        rightMax[n - 1] = height[n - 1];
        for (int i = 1, j = n - 2; j >= 0 && i < n; i++, j--) {
            leftMax[i] = Math.max(leftMax[i - 1], height[i]);
            rightMax[j] = Math.max(rightMax[j + 1], height[j]);
        }
        int res = 0;
        for (int i = 1; i < n - 1; i++) {
            int water = Math.min(leftMax[i], rightMax[i]) - height[i];
            res += water > 0 ? water : 0;
        }
        return res;
    }
    /**
     * 
     * Monolithic Stack:
     * Time: O(n) Space: O(n)
     * We store index of each elements in this stack. In this stack, all elements to which those indexes
     * in stack point to are in descending order. When we meet the top of that stack is smaller than
     * the current height, we can calculate how many water can the cell in the top of stack trap.
     */
    public int trap(int[] height) {
        Stack<Integer> stack = new Stack<>();
        int res = 0;
        for (int i = 0; i < height.length; i++) {
            while (!stack.isEmpty() && height[stack.peek()] < height[i]) {
                int cur = stack.pop();
                if (stack.isEmpty()) {
                    break;
                }
                res += (Math.min(height[stack.peek()], height[i]) - height[cur]) * (i - stack.peek() - 1);
            }
            stack.push(i);
        }
        return res;
    }
}